File: ios/src/main/kotlin/ui/SessionsViewController.kt
import libs.*
import kotlinx.cinterop.*
import platform.UIKit.*
import platform.Foundation.*
import platform.CoreData.*

@ExportObjCClass
@Suppress("CONFLICTING_OVERLOADS")
class SessionsViewController(aDecoder: NSCoder) :
        UIViewController(aDecoder),
        NSFetchedResultsControllerDelegateProtocol,
        UITableViewDataSourceProtocol,
        UITableViewDelegateProtocol
{
    private companion object {
        private val SEND_ID_ONCE_KEY = "sendIfOnce"
    }

    private val favoritesManager = FavoritesManager()

    private var mode: SessionsListMode = SessionsListMode.ALL

    private var _fetchedResultsController: NSFetchedResultsController? = null

    lateinit var pullToRefresh: UIRefreshControl

    @ObjCOutlet lateinit var tableView: UITableView

    private val fetchedResultsController: NSFetchedResultsController
        get() {
            _fetchedResultsController?.let { return it }
            
            val request = NSFetchRequest(entityName = "KSession")
            request.sortDescriptors = nsArrayOf(
                NSSortDescriptor.sortDescriptorWithKey("endsAt", ascending = true),
                NSSortDescriptor.sortDescriptorWithKey("roomName", ascending = true),
                NSSortDescriptor.sortDescriptorWithKey("id", ascending = true))

            if (mode == SessionsListMode.FAVORITES) {
                val favorites = favoritesManager.getFavoriteItemIds()
                request.predicate = NSPredicate.predicateWithFormat("id IN %@", argumentArray = nsArrayOf(favorites))
            }

            val moc = appDelegate.managedObjectContext
            val fetchedResultsController = NSFetchedResultsController(
                request, managedObjectContext = moc, sectionNameKeyPath = "startsAtDate", cacheName = null)
            fetchedResultsController.delegate = this

            _fetchedResultsController = fetchedResultsController
            return fetchedResultsController
        }

    override fun initWithCoder(aDecoder: NSCoder) = initBy(SessionsViewController(aDecoder))

    @ObjCAction
    fun tabSelected(sender: ObjCObject?) {
        val segmentedControl = sender.uncheckedCast<UISegmentedControl>()
        
        mode = if (segmentedControl.selectedSegmentIndex == 0L) SessionsListMode.ALL else SessionsListMode.FAVORITES
        updateResults()
    }

    override fun debugDescription() = "SessionsViewController"

    override fun viewDidLoad() {
        pullToRefresh = UIRefreshControl()
        pullToRefresh.addTarget(this, 
            action = NSSelectorFromString("refreshSessions:"),
             forControlEvents = UIControlEventValueChanged)
        tableView.backgroundView = pullToRefresh

        val moc = appDelegate.managedObjectContext
        val loadSessions = moc.countForFetchRequest(NSFetchRequest(entityName = "KSession"), error = null) <= 0

        if (loadSessions) {
            refreshSessions(this, showProgressPopup = true)
        } else {
            refreshFavorites()
            refreshVotes()
        }

        registerUuid()
    }

    override fun viewWillAppear(animated: Boolean) {
        updateResults()
    }

    private fun registerUuid() {
        // We have to register uuid each time we enter our app, cause the user db may be reset on server
        KonfService(errorHandler = { log(it.localizedDescription) }).registerUser(appDelegate.userUuid)
    }

    @ObjCAction
    fun refreshSessions(sender: ObjCObject?) {
        refreshSessions(sender, showProgressPopup = false)
    }

    private fun refreshSessions(sender: ObjCObject?, showProgressPopup: Boolean) {
        val progressPopup = if (showProgressPopup) showIndeterminateProgress("Loading sessions…") else null

        fun hideProgress() {
            progressPopup?.hideAnimated(true)
            pullToRefresh.endRefreshing()
        }

        val service = KonfService(errorHandler = createErrorHandler("Unable to load sessions.") {
            hideProgress()
        })

        KonfLoader(service).updateSessions {
            hideProgress()
            updateResults()
        }
    }

    private fun refreshFavorites() {
        val service = KonfService(errorHandler = { log(it.localizedDescription) })
        KonfLoader(service).updateFavorites {
            if (mode == SessionsListMode.FAVORITES) {
                updateResults()
            }
        }
    }

    private fun refreshVotes() {
        val service = KonfService(errorHandler = { log(it.localizedDescription) })
        KonfLoader(service).updateVotes()
    }

    private fun updateResults() {
        _fetchedResultsController = null

        try {
            nsTry { errorPtr -> fetchedResultsController.performFetch(error = errorPtr) }
            tableView.reloadData()
        } catch (e: NSErrorException) {
            showPopupText("Unable to load sessions.")
        }
    }

    override fun prepareForSegue(segue: UIStoryboardSegue, sender: ObjCObject?) {
        super.prepareForSegue(segue, sender = sender)

        if (segue.identifier == "ShowSession" && sender != null) {
            val selectedCell = sender.uncheckedCast<SessionsTableViewCell>()
            val selectedPath = tableView.indexPathForCell(selectedCell) ?: return

            val sessionViewController = segue.destinationViewController.uncheckedCast<SessionViewController>()
            sessionViewController.session = fetchedResultsController.objectAtIndexPath(selectedPath)!!.uncheckedCast<KSession>()
        }
    }

    override fun numberOfSectionsInTableView(tableView: UITableView): Long {
        return fetchedResultsController.sections?.count ?: 0
    }

    override fun tableView(tableView: UITableView, didSelectRowAtIndexPath: NSIndexPath) {
        tableView.deselectRowAtIndexPath(didSelectRowAtIndexPath, animated = false)
    }

    override fun tableView(tableView: UITableView, numberOfRowsInSection: Long): Long {
        val sections = fetchedResultsController.sections ?: return 0
        return sections[numberOfRowsInSection].uncheckedCast<NSFetchedResultsSectionInfoProtocol>().numberOfObjects
    }

    override fun tableView(tableView: UITableView, cellForRowAtIndexPath: NSIndexPath): UITableViewCell {
        val cell = tableView.dequeueReusableCellWithIdentifier(
                "Session", forIndexPath = cellForRowAtIndexPath).uncheckedCast<SessionsTableViewCell>()
        val session = fetchedResultsController.objectAtIndexPath(cellForRowAtIndexPath)?.uncheckedCast<KSession>()
        if (session != null) {    
            cell.setup(session)
        }
        return cell
    }
    
    override fun tableView(tableView: UITableView, titleForHeaderInSection: Long): String? {
        val sections = fetchedResultsController.sections ?: return null
        val sectionInfo = sections[titleForHeaderInSection]?.uncheckedCast<NSFetchedResultsSectionInfoProtocol>()
        val session = sectionInfo?.objects?.firstObject?.uncheckedCast<KSession>()
        return renderWeekdayTime(session?.startsAtDate)
    }
}

private enum class SessionsListMode {
    ALL, FAVORITES
}

@ExportObjCClass
class SessionsTableViewCell(aDecoder: NSCoder) : UITableViewCell(aDecoder) {
    @ObjCOutlet lateinit var titleLabel: UILabel
    @ObjCOutlet lateinit var subtitleLabel: UILabel

    override fun initWithCoder(aDecoder: NSCoder) = initBy(SessionsTableViewCell(aDecoder))

    fun setup(session: KSession) {
        titleLabel.text = session.title
        subtitleLabel.text = session.subtitle
    }
}

@ExportObjCClass
class BreakTableViewCell(aDecoder: NSCoder) : UITableViewCell(aDecoder) {
    @ObjCOutlet lateinit var titleLabel: UILabel

    override fun initWithCoder(aDecoder: NSCoder) = initBy(BreakTableViewCell(aDecoder))

    init {
        backgroundColor = UIColor.colorWithPatternImage(UIImage.imageNamed("striped_bg")!!)
    }

    fun setup(session: KSession) {
        titleLabel.text = session.title
    }
}

File: konf-swift/konfswift/ui/SessionsViewController.swift
import UIKit


class SessionsViewController: UITableViewController {
    private static let SEND_ID_ONCE_KEY = "send_uuid_once"
    private lazy var konfService = AppDelegate.me.konfService
    private var mode: KSFKonfServiceSessionsListMode = .all
    private var sessionsTableData: [[KSFSession]] = []

    @IBOutlet weak var pullToRefresh: UIRefreshControl!

    @IBAction func tabSelected(_ sender: Any) {
        guard let segmentedControl = sender as? UISegmentedControl else { return }
        self.mode = (segmentedControl.selectedSegmentIndex == 0) ? .all : .favorites

        self.updateTableContent()
        if let tableView = self.tableView, sessionsTableData.count > 0 {
            tableView.scrollToRow(at: IndexPath(item: 0, section: 0), at: .top, animated: true)
        }
    }

    override func viewDidLoad() {
        if (mode == .all) {
            if (!sendUuidIfNeeded()) { self.refreshSessions(self) }
        } else {
            self.refreshFavorites()
            self.refreshVotes()
        }
    }
    
    private func sendUuidIfNeeded() -> Bool {
        let userDefaults = UserDefaults.standard
        guard !userDefaults.bool(forKey: SessionsViewController.SEND_ID_ONCE_KEY) else { return false }
        
        konfService.register().then(block: { (result) -> KSFStdlibUnit in
            self.refreshSessions(self)
            
            userDefaults.set(result as! Bool, forKey: SessionsViewController.SEND_ID_ONCE_KEY)
            return KUnit
        })
        return true
    }
    
    override func viewWillAppear(_ animated: Bool) {
        self.updateTableContent()
    }

    @IBAction func refreshSessions(_ sender: Any) {
        konfService.refresh().then { (result) -> StdlibUnit in
            self.pullToRefresh?.endRefreshing()
            self.updateTableContent()
            return KUnit
        }.catch { (error) -> StdlibUnit in
            self.showPopupText(title: "Failed to refresh")
            return KUnit
        }
    }

    private func refreshFavorites() {
        konfService.refresh().then { (result) -> KSFStdlibUnit in
            if (self.mode == .favorites) {
                self.updateTableContent()
            }
            return KUnit
        }.catch { (error) -> KSFStdlibUnit in
            return KUnit
        }
    }

    private func refreshVotes() {
        konfService.refresh()
    }
    
    /**
     * Prepare TableView state
     */
    
    private func updateTableContent() {
        switch self.mode {
        case .all:
            fillDataWith(sessions: konfService.sessions)
            break
        case .favorites:
            fillDataWith(sessions: konfService.sessions.filter({ (session) -> Bool in
                return konfService.isFavorite(session: session)
            }))
            break
        default:
            break
        }

        self.tableView?.reloadData()        
    }
    
    private func fillDataWith(sessions: [KSFSession]) {
        let sortedSessions = sessions.sorted(by: { (left, right) -> Bool in
            let byComparator = left.compareTo(other: right)
            if byComparator != 0 { return byComparator < 0 }
            if left.roomId != right.roomId { return left.roomId!.compare(right.roomId!).rawValue > 0 }
            if left.id != right.id { return left.id!.compare(right.id!).rawValue > 0 }
            return false
        })

        sessionsTableData = []
        sortedSessions.forEach({ (session) in
            if sessionsTableData.count == 0 ||
                sessionsTableData.last!.first!.startsAt!.compareTo(otherDate: session.startsAt!) != 0 {
                sessionsTableData.append([session]);
                return
            }
            
            sessionsTableData[sessionsTableData.count - 1].append(session)
        })
    }

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        super.prepare(for: segue, sender: sender)

        switch (segue.identifier ?? "") {
            case "ShowSession":
                guard
                    let selectedCell = sender as? SessionsTableViewCell,
                    let selectedPath = tableView?.indexPath(for: selectedCell)
                else { return }

                let sessionViewController = segue.destination as! SessionViewController
                let bucket = selectedPath.section
                let row = selectedPath.row
                if (sessionsTableData.count <= bucket || sessionsTableData[bucket].count <= row) { return }
                sessionViewController.session = sessionsTableData[bucket][row]
            default: break
        }
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        return sessionsTableData.count
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if sessionsTableData.count <= section { return 0 }
        return sessionsTableData[section].count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Session", for: indexPath) as! SessionsTableViewCell
        let bucket = indexPath.section
        let row = indexPath.row
        if sessionsTableData.count <= bucket || sessionsTableData[bucket].count <= row { return cell }
        cell.setup(for: sessionsTableData[bucket][row])
        return cell
    }
}

class SessionsTableViewCell : UITableViewCell {
    @IBOutlet private weak var dateLabel: UILabel!
    @IBOutlet private weak var titleLabel: UILabel!
    @IBOutlet private weak var icon: UIImageView!

    func setup(for session: KSFSession) {
        titleLabel.text = session.title
        dateLabel.text = KSFUtil.renderDate(date: session.startsAt!)
    }
}
